File: apps/go/design-pattern/builder/builder.go
// Package builder demonstrates the Builder pattern in Go.
// We use method chaining with a builder struct that returns (Computer, error).
package builder

import (
	"errors"
	"fmt"
	"strings"
)

// Drive represents a storage drive.
type Drive struct {
	Type     string // "hdd" or "ssd"
	Capacity int    // in GB
}

func (d Drive) String() string {
	return fmt.Sprintf("%s %dGB", d.Type, d.Capacity)
}

// Motherboard represents a motherboard.
type Motherboard struct {
	Manufacturer string
}

func (m Motherboard) String() string {
	return m.Manufacturer
}

// Computer is the product built by the builder.
type Computer struct {
	Display     string
	Motherboard Motherboard
	Drives      []Drive
	Memory      int // in GB
}

// Describe returns a human-readable description.
func (c *Computer) Describe() string {
	var parts []string
	parts = append(parts, fmt.Sprintf("ディスプレイ: %s", c.Display))
	parts = append(parts, fmt.Sprintf("マザーボード: %s", c.Motherboard))
	parts = append(parts, fmt.Sprintf("メモリ: %dGB", c.Memory))
	for _, d := range c.Drives {
		parts = append(parts, fmt.Sprintf("ドライブ: %s", d))
	}
	return strings.Join(parts, "\n")
}

// ComputerBuilder builds a Computer step by step using method chaining.
type ComputerBuilder struct {
	display     string
	motherboard Motherboard
	drives      []Drive
	memory      int
}

// NewComputerBuilder creates a new builder.
func NewComputerBuilder() *ComputerBuilder {
	return &ComputerBuilder{}
}

// SetDisplay sets the display.
func (b *ComputerBuilder) SetDisplay(display string) *ComputerBuilder {
	b.display = display
	return b
}

// SetMotherboard sets the motherboard.
func (b *ComputerBuilder) SetMotherboard(manufacturer string) *ComputerBuilder {
	b.motherboard = Motherboard{Manufacturer: manufacturer}
	return b
}

// AddDrive adds a drive.
func (b *ComputerBuilder) AddDrive(driveType string, capacity int) *ComputerBuilder {
	b.drives = append(b.drives, Drive{Type: driveType, Capacity: capacity})
	return b
}

// SetMemory sets the memory size.
func (b *ComputerBuilder) SetMemory(gb int) *ComputerBuilder {
	b.memory = gb
	return b
}

// Build creates the Computer, returning an error if required fields are missing.
func (b *ComputerBuilder) Build() (*Computer, error) {
	if b.display == "" {
		return nil, errors.New("ディスプレイは必須です")
	}
	if b.motherboard.Manufacturer == "" {
		return nil, errors.New("マザーボードは必須です")
	}
	if b.memory <= 0 {
		return nil, errors.New("メモリは 1GB 以上必須です")
	}

	return &Computer{
		Display:     b.display,
		Motherboard: b.motherboard,
		Drives:      b.drives,
		Memory:      b.memory,
	}, nil
}

// DesktopBuilder returns a builder pre-configured for a desktop.
func DesktopBuilder() *ComputerBuilder {
	return NewComputerBuilder().
		SetDisplay("27インチ 4K").
		SetMotherboard("ASUS").
		AddDrive("ssd", 512).
		AddDrive("hdd", 2000).
		SetMemory(32)
}

// LaptopBuilder returns a builder pre-configured for a laptop.
func LaptopBuilder() *ComputerBuilder {
	return NewComputerBuilder().
		SetDisplay("15インチ FHD").
		SetMotherboard("Intel").
		AddDrive("ssd", 256).
		SetMemory(16)
}

File: apps/php/design-pattern/src/Builder/Builder.php
<?php

declare(strict_types=1);

namespace DesignPattern\Builder;

class Drive
{
    private string $type;
    private int $sizeGb;

    public function __construct(string $type, int $sizeGb)
    {
        $this->type = $type;
        $this->sizeGb = $sizeGb;
    }

    public function getType(): string
    {
        return $this->type;
    }

    public function getSizeGb(): int
    {
        return $this->sizeGb;
    }
}

class Motherboard
{
    private string $model;

    public function __construct(string $model)
    {
        $this->model = $model;
    }

    public function getModel(): string
    {
        return $this->model;
    }
}

class Computer
{
    private string $display;
    private Motherboard $motherboard;
    /** @var Drive[] */
    private array $drives;
    private int $memoryGb;

    /**
     * @param Drive[] $drives
     */
    public function __construct(
        string $display,
        Motherboard $motherboard,
        array $drives,
        int $memoryGb
    ) {
        $this->display = $display;
        $this->motherboard = $motherboard;
        $this->drives = $drives;
        $this->memoryGb = $memoryGb;
    }

    public function getDisplay(): string
    {
        return $this->display;
    }

    public function getMotherboard(): Motherboard
    {
        return $this->motherboard;
    }

    /** @return Drive[] */
    public function getDrives(): array
    {
        return $this->drives;
    }

    public function getMemoryGb(): int
    {
        return $this->memoryGb;
    }

    public function describe(): string
    {
        $driveInfo = implode(
            ', ',
            array_map(fn(Drive $d) => "{$d->getType()} {$d->getSizeGb()}GB", $this->drives)
        );
        return "Computer: {$this->display}, {$this->motherboard->getModel()}, {$this->memoryGb}GB RAM, [{$driveInfo}]";
    }
}

abstract class ComputerBuilder
{
    protected ?string $display = null;
    protected ?Motherboard $motherboard = null;
    /** @var Drive[] */
    protected array $drives = [];
    protected ?int $memoryGb = null;

    public function setDisplay(string $display): static
    {
        $this->display = $display;
        return $this;
    }

    public function setMotherboard(Motherboard $motherboard): static
    {
        $this->motherboard = $motherboard;
        return $this;
    }

    public function addDrive(Drive $drive): static
    {
        $this->drives[] = $drive;
        return $this;
    }

    public function setMemoryGb(int $memoryGb): static
    {
        $this->memoryGb = $memoryGb;
        return $this;
    }

    abstract protected function validate(): void;

    public function build(): Computer
    {
        $this->validate();
        return new Computer(
            $this->display,
            $this->motherboard,
            $this->drives,
            $this->memoryGb
        );
    }

    public function reset(): static
    {
        $this->display = null;
        $this->motherboard = null;
        $this->drives = [];
        $this->memoryGb = null;
        return $this;
    }
}

class DesktopBuilder extends ComputerBuilder
{
    protected function validate(): void
    {
        if ($this->display === null) {
            throw new \RuntimeException('ディスプレイが設定されていません');
        }
        if ($this->motherboard === null) {
            throw new \RuntimeException('マザーボードが設定されていません');
        }
        if ($this->memoryGb === null) {
            throw new \RuntimeException('メモリが設定されていません');
        }
        if (empty($this->drives)) {
            throw new \RuntimeException('ドライブが1つも追加されていません');
        }
    }
}

class LaptopBuilder extends ComputerBuilder
{
    protected function validate(): void
    {
        if ($this->display === null) {
            throw new \RuntimeException('ディスプレイが設定されていません');
        }
        if ($this->motherboard === null) {
            throw new \RuntimeException('マザーボードが設定されていません');
        }
        if ($this->memoryGb === null) {
            throw new \RuntimeException('メモリが設定されていません');
        }
        if (count($this->drives) > 2) {
            throw new \RuntimeException('ラップトップのドライブは2つまでです');
        }
    }
}
